### PR TITLE
Handle web socket notifications on patient dashboard & archive

### PR DESCRIPTION
**Pull request by nmajor25**
Shortcut Story ID: [sc-59837]

- [x] Handle ws notifications on patient dashboard
- [x] Handle ws notifications on patient archive
- [x] Use `NIL_UUID` for action.flow filter
- [x] Manually test functionality/edge cases in local dev environment
- [x] Add dashboard cypress tests
- [x] Add archive cypress tests
- [x] Fix failing tests - web socket messages intermittently not coming through at beginning of test
- [x] Rebase branch & use new service request implemented in https://github.com/RoundingWell/care-ops-frontend/pull/1417

## Summary by CodeRabbit

- **New Features**
	- Enabled real-time updates for patient data on the archive and dashboard, ensuring that changes to patient flows and actions are reflected live.
	- Improved the visual handling of completed items with smooth fade-out effects for a better user experience.
- **Tests**
	- Expanded test coverage to validate that real-time notifications update the UI consistently for both flows and actions.
	- Added new test specifications for socket notifications related to flows and actions in both the patient archive and dashboard contexts.

### CODE REVIEW

File: src/js/apps/patients/patient/dashboard/dashboard_app.js [hunk 1]
```diff
@@ -25,21 +25,68 @@ export default App.extend({
...
+      Radio.request('entities', 'fetch:actions:collection:byPatient', { patientId: patient.id, filter: this.filter }),
+      Radio.request('entities', 'fetch:flows:collection:byPatient', { patientId: patient.id, filter: this.filter }),
     ];
   },
 
   onStart(options, actions, flows) {
-    this.collection = new Backbone.Collection([...actions.models, ...flows.models]);
+    // remove these when BE api endpoint `filter[state]=` is fixed
+    const doneActions = actions.filter(action => !action.isDone());
+    const doneFlows = flows.filter(flow => !flow.isDone());
```
Comment (nmajor25) [inline src/js/apps/patients/patient/dashboard/dashboard_app.js]: Without this, we'll need to rethink how we determine if a model needs to be fetched or not.

We currently do a `if (this.collection.get(model))`. But without this filtering ☝️, that check won't work in all scenarios.

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 2]
```diff
@@ -13,15 +13,63 @@ export default App.extend({
...
   },
+  subscribe() {
+    const channel = Radio.channel('ws');
+
+    channel.request('subscribe', this.collection.models, {
+      filters: { actions: this.filter, flows: this.filter },
+    });
+
+    this.listenTo(channel, 'message', ({ category, resource, payload }) => {
+      const modelResource = category === 'ResourceCreated' ? payload.resource : resource;
```
Comment (nmajor25) [inline src/js/apps/patients/patient/archive/archive_app.js]: This is technically possible? For a flow or action to be created with a `Done` state? Does that ever happen in the wild? But we should probably handle it either way, right?

File: src/js/views/patients/patient/dashboard/dashboard_views.js [hunk 3]
```diff
@@ -200,7 +200,7 @@ const FlowItemView = View.extend({
   modelEvents: {
     'change:_owner': 'render',
   },
-  behaviors: [RowBehavior],
+  behaviors: [RowBehavior, DoneBehavior],
```
Comment (nmajor25) [inline src/js/views/patients/patient/dashboard/dashboard_views.js:203]: `DoneBehavior` is when the `model.state` changes to `isDone === true`. The action/flow is removed from the list with a fade-out animation.

This was never added for flow list items because it wasn't possible for that to change before. The flow state button is disabled. But now it can change via web socket notification.

File: src/js/apps/patients/patient/dashboard/dashboard_app.js [hunk 4]
```diff
@@ -25,21 +25,68 @@ export default App.extend({
...
+
+      if (this.collection.get(model) || model.isFetching()) return;
+
+      if (modelResource.type === 'flows') {
+        this._fetchFlow(model);
+
+        return;
+      }
+
+      this._fetchAction(model);
```
Comment (nmajor25) [inline src/js/apps/patients/patient/dashboard/dashboard_app.js]: Maybe we can consolidate this to one `this._fetchModel(model)` via `Radio.request('entities', 'fetch:${ type }:model', model.id);`...

File: src/js/apps/patients/patient/dashboard/dashboard_app.js (hunk 1 above)
Comment (paulfalgout) [reply src/js/apps/patients/patient/dashboard/dashboard_app.js]: Why? If it is already in the collection you should never need to fetch it because you are maintaining the data with notifications even if it is not visible

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 5]
```diff
@@ -13,15 +13,63 @@ export default App.extend({
...
-      Radio.request('entities', 'fetch:actions:collection:byPatient', { patientId: patient.id, filter }),
-      Radio.request('entities', 'fetch:flows:collection:byPatient', { patientId: patient.id, filter }),
+      Radio.request('entities', 'fetch:actions:collection:byPatient', { patientId: patient.id, filter: this.filter }),
+      Radio.request('entities', 'fetch:flows:collection:byPatient', { patientId: patient.id, filter: this.filter }),
     ];
   },
   onStart({ patient }, actions, flows) {
-    this.collection = new Backbone.Collection([...actions.models, ...flows.models]);
+    // remove these when BE api endpoint `filter[state]=` is fixed
+    const doneActions = actions.filter(action => action.isDone());
```
Comment (paulfalgout) [inline src/js/apps/patients/patient/archive/archive_app.js]: I don't think we need to do this, the view is handling the filtering enough

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 6]
```diff
@@ -13,15 +13,63 @@ export default App.extend({
...
+
+    this.subscribe();
+
     this.showChildView('content', new ListView({ collection: this.collection }));
   },
+  subscribe() {
+    const channel = Radio.channel('ws');
+
+    channel.request('subscribe', this.collection.models, {
+      filters: { actions: this.filter, flows: this.filter },
```
Comment (paulfalgout) [inline src/js/apps/patients/patient/archive/archive_app.js]: I think both of these filters must include `patientId`

File: src/js/apps/patients/patient/archive/archive_app.js (hunk 2 above)
Comment (paulfalgout) [reply src/js/apps/patients/patient/archive/archive_app.js]: Well we can probably avoid it for now because it probably doesn't happen and we won't need this after the BE change to `ResourceCreated`.   Doing this now just mean we have to test for it.

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 7]
```diff
@@ -13,15 +13,63 @@ export default App.extend({
...
+    });
+
+    this.listenTo(channel, 'message', ({ category, resource, payload }) => {
+      const modelResource = category === 'ResourceCreated' ? payload.resource : resource;
+      const model = Radio.request('entities', 'get:store', modelResource);
+
+      if (this.collection.get(model) || model.isFetching()) return;
+
+      if (modelResource.type === 'flows') {
```
Comment (paulfalgout) [inline src/js/apps/patients/patient/archive/archive_app.js]: yeah I'm perfectly fine combining these with `fetch:${ model.type }:model`   we can probably recombine these methods now since they're simpler than the worklist

Comment (paulfalgout) [reply src/js/apps/patients/patient/archive/archive_app.js]: in fact..  I think we will _always_ have the context needed that this can be
```js
const model = Radio.request('entities', 'get:store', modelResource);

if (this.collection.get(model) || model.isFetching()) return;
model.fetch().then(() => { 
  this.collection.add(model);

  Radio.request('ws', 'add', model);
});
```

unless I'm missing something?

File: src/js/views/patients/patient/dashboard/dashboard_views.js (hunk 3 above)
Comment (paulfalgout) [reply src/js/views/patients/patient/dashboard/dashboard_views.js:203]: this is nice how easy that is to implement

**Review by paulfalgout — CHANGES_REQUESTED** [review]

File: src/js/apps/patients/patient/dashboard/dashboard_app.js (hunk 1 above)
Comment (nmajor25) [reply src/js/apps/patients/patient/dashboard/dashboard_app.js]: Ya I agree, don't think it's needed.

File: src/js/apps/patients/patient/archive/archive_app.js (hunk 6 above)
Comment (nmajor25) [reply src/js/apps/patients/patient/archive/archive_app.js]: That seems right. 

But that will subscribe us to each flow's child actions. So when we receive a ws message for one of those, how will we know if it's an action that should be fetched or not? We'd have to check if that action belonged to a flow, but we won't know that until after we fetch it, right?

Or maybe I'm overthinking it? I need to investigate it more next week.

Comment (nmajor25) [reply src/js/apps/patients/patient/archive/archive_app.js]: I think we'll have to check if the action has a parent flow after fetching it. And if it does, don't add it to the collection.

 I don't think we can determine that until after the action has been fetched.

```js
model.fetch().then(() => {
  if (model.type === 'patient-actions' && model.getFlow()) return;

  this.collection.add(model);

  Radio.request('ws', 'add', model);
});
```

File: src/js/apps/patients/patient/archive/archive_app.js (hunk 7 above)
Comment (nmajor25) [reply src/js/apps/patients/patient/archive/archive_app.js]: Yup that works 👍

File: src/js/apps/patients/patient/archive/archive_app.js (hunk 6 above)
Comment (paulfalgout) [reply src/js/apps/patients/patient/archive/archive_app.js]: @DrechslerDerek we need to have a way to indicate we want to subscribe to actions for a patient that _aren't_ related to a flow.. ideas?

Comment (shadowhand) [reply src/js/apps/patients/patient/archive/archive_app.js]: @paulfalgout I wonder if we need to have two different `types` for patient/flow actions? If we went that route, we might have `patient-actions` and `flow-actions` and then `actions` is a union of them?

Comment (shadowhand) [reply src/js/apps/patients/patient/archive/archive_app.js]: Actually, I think it should be `actions-of-flow` and `actions-of-patient`. Then sorting places them together so there is a visual relationship.

Comment (paulfalgout) [reply src/js/apps/patients/patient/archive/archive_app.js]: Hmm that's a pretty big change and there's been some discussion about decoupling actions from patients at some point, so I'd personally want to avoid it.  Really I wish the current type was just `actions`   I'm wondering if we can specify `filters.actions.flow = null`, or maybe `filters.patientActions` instead of `filters.actions`, to indicate we want patient actions that aren't associated with a flow..  then if there's some reason to split the types we can do so later?

Comment (shadowhand) [reply src/js/apps/patients/patient/archive/archive_app.js]: @paulfalgout if that's not low hanging fruit, then shelve it for now. We can certainly make the current type `actions` with minimal effort, it just requires coordination, right?

In any case... regarding the filter. 😅 I don't like `filter.actions.flow = null` (because PHP can't differentiate between `null` and `undefined` like JS can), but I think `filters.actions.only = 'patient'` (implied enum `patient|flow`) would work?

Comment (shadowhand) [reply src/js/apps/patients/patient/archive/archive_app.js]: But I am also open to `filters.patientActions`. Let's see what makes more sense to @DrechslerDerek.

Comment (paulfalgout) [reply src/js/apps/patients/patient/archive/archive_app.js]: Worth mentioning we'd never use `only = 'flow'` as for that we just set flow to a particular id.  this is just an issue of filtering out flow actions for a patient filter

Comment (DrechslerDerek) [reply src/js/apps/patients/patient/archive/archive_app.js]: I think there could be 2 easy fixes, the first one would be having a flag that says, no flow or flow = null or something, that would be an extra scan condition for dynamo db that would be added dynamically if this flag was set. The other option I see is to do some trick with the current flow filter, where you explicitly pass like 
```
{
    filters: {
        actions: {
            flows: 'null' <or maybe null (at some point in BE this will be a string anyway)>
        }
    }
}
```

which you could actually do right now and it would be a one line change to fix the `$action->flowId ??`. I might be missing how this would cause other issues, and it is more of a bandaid/nuanced solution, but we could do it easily.

Comment (shadowhand) [reply src/js/apps/patients/patient/archive/archive_app.js]: If we're going to do some shenanigans like that, it should use the constant value UUID::NIL.

Comment (DrechslerDerek) [reply src/js/apps/patients/patient/archive/archive_app.js]: Now I don't think this will work, we need to add another DDB `and` or another filter, meaning the flow on the action is dynamic.

File: test/integration/patients/patient/archive.js [hunk 8]
```diff
@@ -652,7 +655,10 @@ context('patient archive page', function() {
       })
       .visitOnClock(`/patient/archive/${ testPatient.id }`, { now: testTs() })
       .wait('@routePatient')
-      .wait('@routePatientActions');
+      .wait('@routePatientActions')
+      .wait('@routePatientFlows');
+
+    cy.wait(200);
```
Comment (nmajor25) [inline test/integration/patients/patient/archive.js:661]: Pushing this up to see if it fixes things when cypress runs in circleci. 

Something is running out of sync. I think it's receiving the first ws notification before the subscription is created. So the app isn't ready to handle it yet. But only sometimes. When I run them locally, it's 50/50 if the test passes or not.

Is there something we're not `.wait()`ing for properly? I can't tell yet.

File: test/integration/patients/patient/archive.js [hunk 9]
```diff
@@ -463,7 +463,9 @@ context('patient archive page', function() {
       })
       .visitOnClock(`/patient/archive/${ testPatient.id }`, { now: testTs() })
       .wait('@routePatient')
-      .wait('@routePatientFlows');
+      .wait('@routePatientActions')
+      .wait('@routePatientFlows')
+      .wait(200); // for initial ws connection to be created
```
Comment (nmajor25) [inline test/integration/patients/patient/archive.js]: Without this `cy.wait()`, these tests fail ~50% of the time. I think what's happening is the `mockWs()` command isn't always finished before we send the first ws message in the test via `cy.sendWs()`.

Adding a `cy.wait(200)` seems to fix the issue, where the test passes 100% of the time. I don't think this does much harm. 

But I would prefer to not do this if there was another way to fix it. The `mockWs` would have to be reconfigured somehow, to somehow let us know when the initial socket connection is ready.

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 10]
```diff
@@ -1,19 +1,22 @@
...
   onBeforeStart({ patient }) {
+    this.patient = patient;
     this.showView(new LayoutView({ model: patient }));
     this.getRegion('content').startPreloader();
   },
   beforeStart({ patient }) {
     const currentWorkspace = Radio.request('workspace', 'current');
-    const states = currentWorkspace.getStates();
-    const filter = { states: states.groupByDone().done.getFilterIds() };
+    this.states = currentWorkspace.getStates();
```
Comment (paulfalgout) [inline src/js/apps/patients/patient/archive/archive_app.js]: I think it'd be better to set this in the `onBeforeStart` and not attach things in the `beforeStart`.. and you can probably save the `states` since it's reused in the `subscribe`

**Review by paulfalgout — CHANGES_REQUESTED** [review]

File: src/js/apps/patients/patient/archive/archive_app.js (hunk 10 above)
Comment (nmajor25) [reply src/js/apps/patients/patient/archive/archive_app.js]: Like this: https://github.com/RoundingWell/care-ops-frontend/pull/1400/commits/af63bdd3d766db9857e159ffa369e8e08fe22252?

**Comment by paulfalgout** [comment]
@nmajor25 I think this could be rebased and modified with the new service request and maybe it'd work?

**Comment by nmajor25** [comment]
@paulfalgout think so, I'll give it a try 👍

File: src/js/apps/patients/patient/archive/archive_app.js [hunk 11]
```diff
@@ -22,6 +27,21 @@ export default App.extend({
...
+    const filters = {
+      states: this.states,
+      patient: this.patient.id,
+    };
+
+    Radio.request('ws', 'subscribe', this.collection.models, {
+      filters: { actions: { ...filters, flow: NIL_UUID }, flows: filters },
+    });
+    Radio.request('ws', 'manage:add', this, this.collection, 'flows');
+    Radio.request('ws', 'manage:add', this, this.collection, 'patient-actions');
```
Comment (nmajor25) [inline src/js/apps/patients/patient/archive/archive_app.js:45]: Two separate service requests here, one for flows and another for actions.

Comment (nmajor25) [reply src/js/apps/patients/patient/archive/archive_app.js:45]: I tested this locally for both patient dashboard and archive, it worked well.

File: test/integration/patients/patient/archive.js (hunk 9 above)
Comment (paulfalgout) [reply test/integration/patients/patient/archive.js]:
```js
    cy
      .get('@wsHandleMessage')
      .should('have.been.calledOnce');
```
will do it.

The issue was that the `send` is happening in Cypress before the `subscribe` happens from the app.. so we have to wait until the subscribe is called.

**Review by paulfalgout — APPROVED** [review]
